compiler/tests/fixtures/demos/uniswap.fe

```diff
@@ -0,0 +1,359 @@
+    # const PERMIT_TYPEHASH: u256 = 0x6e71edae12b1b97f4d1f60370fef10105fa2faae0126114a169c64845d6126c9
+    nonces: map<address, u256>
+
+    event Approval:
+        idx owner: address
+        idx spender: address
+        value: u256
+
+    event Transfer:
+        idx from: address
+        idx to: address
+        value: u256
+
+    pub def __init__():
+        self.DOMAIN_SEPARATOR = (
+            "EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)".keccak256(),
+            self.name.keccak256(),
+            u256(1).keccak256(),
+            chain.id,
+            self.address
+        ).abiEncode().keccak256()
+
+    def _mint(to: address, value: u256):
+        self.total_supply = self.total_supply + value
+        self.balances[to] = self.balances[to] + value
+        emit Transfer(address(0), to, value)
+
+    def _burn(from: address, value: u256):
+        self.balances[from] = self.balances[from] - value
+        self.total_supply = total_supply - value
+        emit Transfer(from, address(0), value)
+
+    def _approve(owner: address, spender: address, value: u256):
+        allowance[owner][spender] = value
+        emit Approval(owner, spender, value)
+
+    def _transfer(from: address, to: address, value: u256):
+        balances[from] = balances[from] - value
+        balances[to] = balances[to] - value
+        emit Transfer(from, to, value)
+
+    pub def approve(spender: address, value: u256) -> bool:
+        self._approve(msg.sender, spender, value)
+        return true
+
+    pub def transfer(to: address, value: u256) -> bool:
+        self._transfer(msg.sender, to, value)
+        return true
+
+    pub def transferFrom(from: address, to: address, value: u256) -> bool:
+        self.allowances[from][msg.sender] = allowances[from][msg.sender] - value
+        self._transfer(from, to, value)
+        return true
+
+    # we may want to pass on this feature for now
+    pub def permit(
+        owner: address,
+        spender: address,
+        value: u256,
+        deadline: u256,
+        v: u8,
+        r: u256,
+        s: u256
+    ):
+       assert deadline >= block.timestamp, "UniswapV2: EXPIRED"
+       digest: u256 = (
+           0x1901,
+           DOMAIN_SEPARATOR,
+           # (PERMIT_TYPEHASH, owner, spender, value, nonces[owner]++, deadline).abiEncode().keccak256()
+       ).abiEncodePacked().keccak256()
+       recovered_address: address = crypto.ecrecover(digest, v, r, s)
+       assert recovered_address != address(0) and recoveredAddress == owner, "UniswapV2: INVALID_SIGNATURE"
+       self._approve(owner, spender, value)
+
+
+contract UniswapV2Pair:
+    # const MINIMUM_LIQUIDITY: u256 = 1000
+    # todo set the value below to keccak256('transfer(address,uint256)')
+    # const SELECTOR: u32 = 0
+
+    factory: address
+    token0: address
+    token1: address
+
+    reserve0: u112
```

Comment:
I haven't seen the code that does packing of the variables or maybe I did but missed it. I am more inclined to use every functionality that we have if support smart packing by the compiler itself then let's use it.

Reply:
The `u112` part is a direct carryover from the solidity implementation. I left it here for the time being just so we don't forget about something that's used to cut gas costs. It should be removed, though, since we don't support integers of size 112. Right now all base types occupy a whole word in storage, so there is no packing going on.

Reply:
That make sense now.